src/PowerShell/Microsoft.WinGet.Configuration.Engine/Commands/AsyncCommand.cs

```diff
@@ -51,62 +41,74 @@ public abstract class AsyncCommand
         {
+            // Passing Debug will make all the message actions to be Inquire. For async operations
+            // and the current queue message implementation this doesn't make sense.
+            // PowerShell will inquire for any message giving the impression that the task is
+            // paused, but the async operation is still running.
+            if (psCmdlet.MyInvocation.BoundParameters.ContainsKey("Debug"))
+            {
+                throw new NotSupportedException(Resources.DebugNotSupported);
+            }
+
             this.PsCmdlet = psCmdlet;
             this.originalThread = Thread.CurrentThread;
-            this.isDebugBounded = this.PsCmdlet.MyInvocation.BoundParameters.ContainsKey("Debug");
-            this.canWriteToStream = canWriteToStream;
             this.cancellationToken = this.source.Token;
         }
 
-        private enum OutputStreamType
+        /// <summary>
+        /// The write stream type.
+        /// </summary>
+        public enum StreamType
         {
+            /// <summary>
+            /// Debug.
+            /// </summary>
             Debug,
+
+            /// <summary>
+            /// Verbose.
+            /// </summary>
             Verbose,
+
+            /// <summary>
+            /// Warning.
+            /// </summary>
             Warning,
+
+            /// <summary>
+            /// Error.
+            /// </summary>
             Error,
+
+            /// <summary>
+            /// Progress.
+            /// </summary>
             Progress,
+
+            /// <summary>
+            /// Object.
+            /// </summary>
+            Object,
+
+            /// <summary>
+            /// Information.
+            /// </summary>
+            Information,
         }
 
         /// <summary>
         /// Gets the base cmdlet.
         /// </summary>
         protected PSCmdlet PsCmdlet { get; private set; }
 
-        /// <summary>
-        /// Gets or sets a value indicating whether if writing to stream is blocked.
-        /// Writing to streams must be blocked for Start-* cmdlets. The Complete-* cmdlet counterpart
-        /// will enable writing to the stream when executed.
-        /// TODO: For now any messages before the Complete-* call get lost. We can add a ConcurrentQueue
-        /// to store message and flush them in or before the Wait call.
-        /// </summary>
-        private bool CanWriteToStream
-        {
-            get
-            {
-                lock (CmdletLock)
-                {
-                    return this.canWriteToStream;
-                }
-            }
-
-            set
-            {
-                lock (CmdletLock)
-                {
-                    this.canWriteToStream = value;
-                }
-            }
-        }
-
         /// <summary>
         /// Cancel this operation.
         /// </summary>
-        public virtual void Cancel()
+        public virtual void Complete()

```

Comment:
Did we lose `Cancel`?  The comment on this function is still for `Cancel`.

Reply:
I used to have a cancel that no one called. I'll bring it back when integrate your cancellation support.